File: jarvis-client/app/src/main/java/com/jarvis/client/voice/VoiceFlow.kt
package com.jarvis.client.voice

import com.jarvis.client.audio.Wav
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sin

/**
 * The voice flow on this phone (docs/JARVIS-API.md section 17):
 * interrupting Jarvis by talking, "One moment." when a tool starts, a small
 * "I heard you" sound, and `waited_ms`. Pure - no Android - so it is tested
 * (`VoiceFlowTest`), and these are the desktop's rules word for word
 * (jarvis-desktop `src/voice-flow.js`): both apps are held to one list of
 * cases, `jarvis-desktop/tests/fixtures/voice-flow-cases.json`.
 *
 * INTERRUPTING BY TALKING - "pause first, decide second" (LiveKit agents'
 * shape; the voice research of 2026-09-24, recommendation 4). While Jarvis
 * speaks and "Interrupt Jarvis while it talks" is on, the hands-free
 * listener hears the room through the echo canceller. When [SpeechRun]
 * has heard [ONSET_MS] of speech, the reply is PAUSED at once
 * ([InterruptFlow.onset]) - not talked over for the two seconds the check
 * needs - and about [CLIP_MS] of that speech, from where it started, goes
 * to the PC as `source=barge_in`. The PC answers one question: was it the
 * owner (or the word "stop")? Yes: the reply stops for good, and a sentence
 * made ahead is thrown away (SpeechAhead). No: it carries on from where it
 * paused. No answer within [WAIT_MAX_MS] of pausing: it carries on too.
 * Nothing on this phone turns the speech into words: [SpeechRun] measures
 * loudness only, the same rule the talk button's pause detector uses, and
 * the PC never transcribes a barge-in clip.
 *
 * The first [GRACE_MS] of a reply are ignored: that is when the echo
 * canceller is still learning the room and Jarvis's own voice most often
 * comes back through the microphone (kyutai unmute waits 3 s for the same
 * reason). "Stop" is not affected: the phone's own stop-word head
 * ([BargeIn]) works from the first word, as before.
 */
object VoiceFlow {
    /** The first part of a reply that talking over is ignored for (echo). */
    const val GRACE_MS = 3_000L

    /** Speech heard before the reply is paused. */
    const val ONSET_MS = 500L

    /** About this much speech goes to the PC: it recognised the owner from 2 s clips, not from 1.2-1.5 s. */
    const val CLIP_MS = 2_000L

    /** After the pause started, this much quiet ends the speech early: what there is goes to the PC. */
    const val END_QUIET_MS = 700L

    /** Paused and no answer from the PC this long: carry on talking. */
    const val WAIT_MAX_MS = 4_000L

    /** Kept from before the speech began, so its first sound is in the clip. */
    const val PREROLL_MS = 300L

    enum class Action { NONE, PAUSE, RESUME, STOP }

    /**
     * The marks that end a question - the PC's `jarvis_speech.QUESTION_MARKS`
     * (the shared cases' `question_marks`): ASCII, full-width, and the Greek
     * question mark U+037E.
     */
    val QUESTION_MARKS: List<Char> = listOf('?', '？', ';')

    private val GREEK = Regex("[\\u0370-\\u03ff\\u1f00-\\u1fff]")

    /**
     * Keep listening after a question (section 17, 5): does [text] - the last
     * sentence Jarvis spoke - end with a question mark? Closing quotes and
     * brackets after it are allowed; a plain ";" counts only in Greek, where
     * Unicode folds the Greek question mark into it. The PC's
     * `ends_with_question`, the same rule.
     */
    fun endsWithQuestion(text: String?): Boolean {
        val t = (text ?: "").trimEnd().trimEnd('"', '\'', '”', '’', ')', ']', '»').trimEnd()
        if (t.isEmpty()) return false
        val last = t.last()
        return last in QUESTION_MARKS || (last == ';' && GREEK.containsMatchIn(t))
    }

    /** Whether a `step` event's `data` says a tool is starting. */
    fun isToolStart(data: JsonElement?): Boolean =
        ((data as? JsonObject)?.get("phase") as? JsonPrimitive)?.takeIf { it.isString }?.content == "tool_started"

    /**
     * How long the end of a clip was quiet, in milliseconds: `waited_ms`,
     * "from the moment the speech detector last heard speech to the moment
     * the clip is sent" (section 17, 2). A 30 ms window is speech when it is
     * louder than three times the quietest window in the clip, within the
     * listener's own bounds ([WakeClip.EndOfSpeech.threshold]) - the desktop's
     * `trailing_quiet` (voice_flow.rs), the same rule. Null when nothing in
     * the clip was loud enough to be speech.
     */
    fun trailingQuietMs(pcm: ShortArray, rate: Int = Wav.SAMPLE_RATE): Long? {
        val window = max(1, rate * 30 / 1000)
        if (pcm.isEmpty()) return null
        val levels = (pcm.indices step window).map { start ->
            val n = min(window, pcm.size - start)
            Wav.rms(pcm.copyOfRange(start, start + n), n)
        }
        val floor = levels.minOrNull() ?: return null
        val bar = WakeClip.EndOfSpeech.threshold(floor)
        val lastLoud = levels.indexOfLast { it >= bar }
        if (lastLoud < 0) return null
        return (levels.size - 1 - lastLoud) * 30L
    }
}

/**
 * Interrupting by talking, as a small state machine - `createInterruptFlow`
 * in the desktop's voice-flow.js. Every call takes the time it happened at
 * ([now], milliseconds on one clock) and says what to do.
 */
class InterruptFlow {
    private var replyStartedAt: Long? = null
    private var pausedAt: Long? = null
    private var asked: Long? = null

    /** The reply made its first sound. */
    @Synchronized
    fun replyStarted(now: Long) {
        if (replyStartedAt == null) replyStartedAt = now
    }

    /** The reply is over, stopped, or a new question was asked. */
    @Synchronized
    fun replyEnded() {
        replyStartedAt = null
        pausedAt = null
        asked = null
    }

    @get:Synchronized
    val paused: Boolean get() = pausedAt != null

    /**
     * Half a second of speech was heard, clip [id]. [allowed]: the switch is
     * on, the PC says it can tell the owner's voice now, and the link is not
     * stale. PAUSE means: pause the reply and have that clip checked.
     */
    @Synchronized
    fun onset(now: Long, id: Long, allowed: Boolean): VoiceFlow.Action {
        val started = replyStartedAt
        if (!allowed || pausedAt != null || started == null) return VoiceFlow.Action.NONE
        if (now - started < VoiceFlow.GRACE_MS) return VoiceFlow.Action.NONE
        pausedAt = now
        asked = id
        return VoiceFlow.Action.PAUSE
    }

    /** The PC's answer for clip [id]: [stop] true when it was the owner or "stop". */
    @Synchronized
    fun verdict(id: Long, stop: Boolean): VoiceFlow.Action {
        if (asked == null || asked != id) return VoiceFlow.Action.NONE
        asked = null
        if (stop) {
            replyStartedAt = null
            pausedAt = null
            return VoiceFlow.Action.STOP
        }
        if (pausedAt == null) return VoiceFlow.Action.NONE
        pausedAt = null
        return VoiceFlow.Action.RESUME
    }

    /** Time passing: paused too long with no answer means carry on. */
    @Synchronized
    fun tick(now: Long): VoiceFlow.Action {
        val at = pausedAt ?: return VoiceFlow.Action.NONE
        if (now - at < VoiceFlow.WAIT_MAX_MS) return VoiceFlow.Action.NONE
        pausedAt = null
        return VoiceFlow.Action.RESUME
    }
}

/** "One moment.", at most once per spoken question - `createMomentFlow` in voice-flow.js. */
class MomentFlow {
    private var active = false
    private var played = false
    private var replied = false
    private var stopped = false

    @Synchronized
    fun turnStarted() {
        active = true
        played = false
        replied = false
        stopped = false
    }

    @Synchronized
    fun turnEnded() {
        active = false
    }

    /** The reply made (or is about to make) its first sound. */
    @Synchronized
    fun replyStarted() {
        replied = true
    }

    /** "Stop", or an interruption that stopped the reply. */
    @Synchronized
    fun stopped() {
        stopped = true
    }

    /** A tool started. [enabled]: the owner's switch and the PC's; [ready]: the clip is here. */
    @Synchronized
    fun toolStarted(enabled: Boolean, ready: Boolean): Boolean {
        if (!active || played || replied || stopped || !enabled || !ready) return false
        played = true
        return true
    }
}

/**
 * Speech over a reply, from per-step loudness only (the talk button's own
 * rule, [WakeClip.EndOfSpeech.threshold]). [step] is fed every 80 ms step
 * the barge-in listener reads, and says when half a second of speech has
 * been heard ([Event.ONSET]) and then when its clip is ready to send
 * ([Event.CLIP_DUE]): [VoiceFlow.CLIP_MS] after the speech began, or once
 * [VoiceFlow.END_QUIET_MS] of quiet ends it sooner. After a clip, a new
 * run begins only after that quiet.
 */
class SpeechRun(initialFloor: Float = 0.005f) {
    enum class Event { NONE, ONSET, CLIP_DUE }

    private var floor = initialFloor

    /** When the current run of speech began, or null in quiet. */
    var startedAt: Long? = null
        private set

    private var voicedMs = 0L
    private var lastVoicedAt = 0L
    private var told = false
    private var sent = false

    fun step(now: Long, level: Float, stepMs: Long): Event {
        val voiced = level >= WakeClip.EndOfSpeech.threshold(floor)
        if (!voiced) floor = WakeClip.EndOfSpeech.updateFloor(floor, level)
        val begun = startedAt
        if (begun == null) {
            if (!voiced) return Event.NONE
            startedAt = now - stepMs
            voicedMs = stepMs
            lastVoicedAt = now
            told = false
            sent = false
            return if (voicedMs >= VoiceFlow.ONSET_MS) onset() else Event.NONE
        }
        if (voiced) {
            voicedMs += stepMs
            lastVoicedAt = now
        }
        val quiet = now - lastVoicedAt
        if (!told) {
            if (voicedMs >= VoiceFlow.ONSET_MS) return onset()
            if (quiet >= VoiceFlow.END_QUIET_MS) startedAt = null // a cough: forgotten
            return Event.NONE
        }
        if (!sent && (now - begun >= VoiceFlow.CLIP_MS || quiet >= VoiceFlow.END_QUIET_MS)) {
            sent = true
            return Event.CLIP_DUE
        }
        if (sent && quiet >= VoiceFlow.END_QUIET_MS) startedAt = null
        return Event.NONE
    }

    private fun onset(): Event {
        told = true
        return Event.ONSET
    }
}

/**
 * Where the owner cut the last spoken answer off (section 17, 6): the
 * sentence they heard last, handed to the NEXT question - typed or spoken -
 * as `interrupted`, so the PC can tell its model that the answer was cut off
 * there. Once only, and only within [keepMs]: a question an hour later is
 * not a reply to the interruption. The desktop's `createCutOff`.
 */
class CutOff(private val keepMs: Long = KEEP_MS) {
    private var said: String? = null
    private var at = 0L

    /** The owner stopped the answer while [sentence] was the last one heard. */
    @Synchronized
    fun cut(sentence: String?, now: Long) {
        val s = sentence?.trim()
        if (s.isNullOrEmpty()) return
        said = s
        at = now
    }

    /** The sentence for the question being sent now, or null - and forgotten either way. */
    @Synchronized
    fun take(now: Long): String? {
        val s = said
        said = null
        return s?.takeIf { now - at <= keepMs }
    }

    companion object {
        /** Two minutes. */
        const val KEEP_MS = 120_000L
    }
}

/** The PC's answer to a barge-in clip. */
data class BargeVerdict(val stop: Boolean, val available: Boolean, val why: String) {
    companion object {
        /** An interruption that could not be checked never stops the reply. */
        val FAILED = BargeVerdict(stop = false, available = true, why = "failed")

        /**
         * `{"stop", "available", "why", ...}`. A route without
         * voice-flow.patch answers the usual utterance reply, whose `stop`
         * means the same (section 17). Anything unreadable is [FAILED].
         */
        fun read(body: JsonObject?): BargeVerdict {
            if (body == null) return FAILED
            fun flag(key: String): Boolean? = (body[key] as? JsonPrimitive)?.takeIf { !it.isString }?.booleanOrNull
            return BargeVerdict(
                stop = flag("stop") ?: false,
                available = flag("available") ?: true,
                why = (body["why"] as? JsonPrimitive)?.takeIf { it.isString }?.contentOrNull.orEmpty(),
            )
        }
    }
}

/** "Say 'One moment' if I'm kept waiting": this phone's own switch, on by default. */
object OneMoment {
    /** The switch's name - the desktop's `MOMENT_NAME`, the words section 17 suggests. */
    const val NAME = "Say \"One moment\" if I'm kept waiting"

    /** The desktop's `describeMoment`, word for word. */
    fun describe(on: Boolean): String = if (on) {
        "On: when Jarvis has to look something up for a spoken question, it says \"One moment.\" " +
            "first, once, in its own voice."
    } else {
        "Off: Jarvis stays quiet until its answer is ready."
    }
}

/**
 * "I heard you": a tiny two-note sound when the owner's turn is cut, made
 * from numbers (no file, no new dependency) - the desktop's `HEARD_SOUND`,
 * the same numbers, so both apps sound alike.
 */
object HeardSound {
    const val RATE = 24_000
    val TONES: List<Pair<Int, Int>> = listOf(660 to 55, 990 to 75) // (hz, ms)
    const val GAP_MS = 25
    const val FADE_MS = 10
    const val GAIN = 0.16

    fun samples(): ShortArray {
        val fade = (RATE * FADE_MS / 1000.0).roundToInt()
        val gap = (RATE * GAP_MS / 1000.0).roundToInt()
        val out = ArrayList<Short>()
        TONES.forEachIndexed { i, (hz, ms) ->
            if (i > 0) repeat(gap) { out.add(0) }
            val n = (RATE * ms / 1000.0).roundToInt()
            for (k in 0 until n) {
                val edge = minOf(1.0, k.toDouble() / fade, (n - 1 - k).toDouble() / fade)
                val v = sin(2 * PI * hz * k / RATE) * GAIN * max(0.0, edge)
                out.add(jsRound(v * 32767).toShort())
            }
        }
        return out.toShortArray()
    }

    /** JavaScript's Math.round (halves go up), so both apps make the same numbers. */
    private fun jsRound(x: Double): Int = kotlin.math.floor(x + 0.5).toInt()

    /** The loudest sample, for the test. */
    fun peak(s: ShortArray): Int = s.maxOfOrNull { abs(it.toInt()) } ?: 0
}

File: jarvis-desktop/src/voice-flow.js
/**
 * The voice flow on this PC (docs/JARVIS-API.md section 17): interrupting
 * Jarvis by talking, "One moment." when a tool starts, and a small "I heard
 * you" sound. The rules here are pure - no Tauri, no audio - so they are
 * tested (tests/voice-flow.mjs), and they are the phone's rules word for
 * word (jarvis-client voice/VoiceFlow.kt): both apps are held to one list of
 * cases, tests/fixtures/voice-flow-cases.json.
 *
 * INTERRUPTING BY TALKING - "pause first, decide second" (LiveKit agents'
 * shape, voice/agent_activity.py; the voice research of 2026-09-24,
 * recommendation 4). While Jarvis speaks and this PC listens for "hey
 * Jarvis", voice.rs tells the Jarvis bar when half a second of speech has
 * been heard (`voice-barge-onset`). The bar PAUSES the reply at once - it
 * does not keep talking over the owner for the two seconds the check needs -
 * and asks voice.rs to send about two seconds of that speech to the PC
 * (`source=barge_in`). The PC answers one question: was it the owner (or
 * the word "stop")? Yes: the reply stops for good, and a sentence already
 * made ahead is thrown away. No: the reply carries on from where it paused.
 * No answer within `WAIT_MAX_MS` of pausing: it carries on too.
 *
 * The first `GRACE_MS` of a reply are ignored: that is when the echo
 * canceller is still learning the room and Jarvis's own voice is most
 * likely to come back through the microphone (kyutai unmute waits 3 s for
 * the same reason). "Stop" is not affected by that: the stop word has its
 * own path (the PC answers `stop: true` to a short clip) and works from the
 * first word.
 *
 * "ONE MOMENT." Played when a tool starts (the `step` event with
 * `tool_started` - the same events the private-answer rule watches), at
 * most once per spoken question, and only before the reply has made any
 * sound: never over the reply, never after "stop". Not on a flat one-second
 * timer, which at today's speeds would fire on almost every spoken turn
 * (section 17, 3).
 *
 * "I HEARD YOU." A tiny two-note sound, made here from numbers (no file, no
 * new dependency), played when the owner's turn has been cut: on letting go
 * of the talk button, and when the PC says "hey Jarvis" was heard. The phone
 * makes the same sound from the same numbers.
 *
 * @module voice-flow
 */

/** The first part of a reply that talking over is ignored for (echo). */
export const GRACE_MS = 3000;
/** Speech heard before the reply is paused (voice.rs `BARGE_ONSET`). */
export const ONSET_MS = 500;
/** About this much speech goes to the PC (voice.rs `BARGE_CLIP`). */
export const CLIP_MS = 2000;
/** Paused and no answer from the PC this long: carry on talking. */
export const WAIT_MAX_MS = 4000;

export const NONE = "none";
export const PAUSE = "pause";
export const RESUME = "resume";
export const STOP = "stop";

/**
 * Interrupting by talking, as a small state machine. Every method takes the
 * time it happened at (`now`, milliseconds on one clock) and returns what
 * to do: `NONE`, `PAUSE`, `RESUME` or `STOP`.
 */
export function createInterruptFlow() {
  let replyStartedAt = null;
  let pausedAt = null;
  let asked = null;
  return {
    /** The reply made its first sound (the first clip started playing). */
    replyStarted(now) {
      if (replyStartedAt === null) replyStartedAt = now;
    },
    /** The reply is over, stopped, or a new question was asked. */
    replyEnded() {
      replyStartedAt = null;
      pausedAt = null;
      asked = null;
    },
    /** Paused, waiting for the PC's answer. */
    paused() {
      return pausedAt !== null;
    },
    /**
     * Half a second of speech was heard, clip `id`. `allowed`: the switch
     * is on, the PC says it can tell the owner's voice now, and the link is
     * not stale. PAUSE means: pause the reply and have that clip checked.
     */
    onset(now, id, allowed) {
      if (!allowed || pausedAt !== null || replyStartedAt === null) return NONE;
      if (now - replyStartedAt < GRACE_MS) return NONE;
      pausedAt = now;
      asked = id;
      return PAUSE;
    },
    /** The PC's answer for clip `id`: `stop` true when it was the owner or "stop". */
    verdict(id, stop) {
      if (asked === null || asked !== id) return NONE;
      asked = null;
      if (stop === true) {
        replyStartedAt = null;
        pausedAt = null;
        return STOP;
      }
      if (pausedAt === null) return NONE;
      pausedAt = null;
      return RESUME;
    },
    /** Time passing: paused too long with no answer means carry on. */
    tick(now) {
      if (pausedAt === null || now - pausedAt < WAIT_MAX_MS) return NONE;
      pausedAt = null;
      return RESUME;
    },
  };
}

/**
 * "One moment.", at most once per spoken question. `toolStarted` says
 * whether to play it now.
 */
export function createMomentFlow() {
  let active = false;
  let played = false;
  let replied = false;
  let stopped = false;
  return {
    /** A spoken question was sent. */
    turnStarted() {
      active = true;
      played = false;
      replied = false;
      stopped = false;
    },
    /** A typed question replaced it, or the turn is over. */
    turnEnded() {
      active = false;
    },
    /** The reply made its first sound. */
    replyStarted() {
      replied = true;
    },
    /** "Stop", or an interruption that stopped the reply. */
    stopped() {
      stopped = true;
    },
    /** A tool started. `enabled`: the owner's switch and the PC's; `ready`: the clip is here. */
    toolStarted(enabled, ready) {
      if (!active || played || replied || stopped || !enabled || !ready) return false;
      played = true;
      return true;
    },
  };
}

/**
 * Where the owner cut the last spoken answer off (section 17, 6): the
 * sentence they heard last, handed to the NEXT question - typed or spoken -
 * as `interrupted` on its user message, so the PC can tell its model the
 * answer was cut off there. Once only, and only within `CUT_OFF_KEEP_MS`.
 * The phone's `CutOff`.
 */
export const CUT_OFF_KEEP_MS = 120000;

export function createCutOff(keepMs = CUT_OFF_KEEP_MS) {
  let said = null;
  let at = 0;
  return {
    /** The owner stopped the answer while `sentence` was the last one heard. */
    cut(sentence, now) {
      const s = typeof sentence === "string" ? sentence.trim() : "";
      if (!s) return;
      said = s;
      at = now;
    },
    /** The sentence for the question being sent now, or null - forgotten either way. */
    take(now) {
      const s = said;
      said = null;
      return s !== null && now - at <= keepMs ? s : null;
    },
  };
}

/** Whether a `step` event's `data` says a tool is starting. */
export function isToolStart(data) {
  return Boolean(data && typeof data === "object" && data.phase === "tool_started");
}

/**
 * What the PC's `/api/voice/status` `flow` block allows, read defensively:
 * an older PC (no `flow`) allows nothing - never send `source=barge_in` to
 * it (section 17: an older PC would transcribe the clip).
 */
export function flowFromStatus(status) {
  const flow = status && typeof status === "object" && status.flow && typeof status.flow === "object"
    ? status.flow
    : null;
  const barge = flow && flow.barge_in && typeof flow.barge_in === "object" ? flow.barge_in : {};
  const moment = flow && flow.moment && typeof flow.moment === "object" ? flow.moment : {};
  const on = Boolean(flow && flow.available === true);
  return {
    bargeIn: on && barge.enabled === true && barge.available === true,
    bargeInWhy: typeof barge.why === "string" ? barge.why : "",
    moment: on && moment.enabled === true,
    momentReady: on && moment.enabled === true && moment.ready === true,
    momentKey: typeof moment.key === "string" ? moment.key : "",
    // 2026-09-25: the PC keeps listening after a question, and reads
    // `interrupted` and keeps it on the PC. Neither is sent to an older PC.
    afterQuestion: on && flow.after_question === true,
    cutOff: on && flow.cut_off === true,
  };
}

/* ── The "One moment." switch - this PC's own, like "Interrupt Jarvis while it talks" ── */

/** The localStorage key. */
export const MOMENT_KEY = "jarvis.voice.oneMoment";

/** The switch's name, the words JARVIS-API section 17 suggests - the phone's too. */
export const MOMENT_NAME = "Say \"One moment\" if I'm kept waiting";

/** The owner's setting, or ON when none was saved (or storage is unreadable). */
export function loadMoment(storage = globalThis.localStorage) {
  try {
    if (storage && storage.getItem(MOMENT_KEY) === "off") return false;
  } catch {
    /* private mode or cleared site data: the default */
  }
  return true;
}

/** Saves the setting. Returns whether it could be saved. */
export function saveMoment(on, storage = globalThis.localStorage) {
  try {
    if (!storage) return false;
    storage.setItem(MOMENT_KEY, on ? "on" : "off");
    return true;
  } catch {
    return false;
  }
}

/** The switch's line in Settings - the phone's `OneMoment.describe`, word for word. */
export function describeMoment(on) {
  // One literal each: the phone's VoiceFlowTest finds these very lines here.
  return on
    ? "On: when Jarvis has to look something up for a spoken question, it says \"One moment.\" first, once, in its own voice."
    : "Off: Jarvis stays quiet until its answer is ready.";
}

/* ── "I heard you" ── */

/** The sound's numbers, the phone's `HeardSound` numbers. */
export const HEARD_SOUND = Object.freeze({
  rate: 24000,
  tones: Object.freeze([Object.freeze({ hz: 660, ms: 55 }), Object.freeze({ hz: 990, ms: 75 })]),
  gapMs: 25,
  fadeMs: 10,
  gain: 0.16,
});

/** The sound as 16-bit samples: each note a sine with short fades, a gap between. */
export function heardSoundSamples(spec = HEARD_SOUND) {
  const rate = spec.rate;
  const fade = Math.round((rate * spec.fadeMs) / 1000);
  const gap = Math.round((rate * spec.gapMs) / 1000);
  const out = [];
  spec.tones.forEach((tone, i) => {
    if (i > 0) for (let k = 0; k < gap; k += 1) out.push(0);
    const n = Math.round((rate * tone.ms) / 1000);
    for (let k = 0; k < n; k += 1) {
      const edge = Math.min(1, k / fade, (n - 1 - k) / fade);
      const v = Math.sin((2 * Math.PI * tone.hz * k) / rate) * spec.gain * Math.max(0, edge);
      out.push(Math.round(v * 32767));
    }
  });
  return Int16Array.from(out);
}

/** A mono 16-bit WAV of `samples` at `rate`, as bytes. */
export function wavBytes(samples, rate) {
  const data = samples.length * 2;
  const buf = new ArrayBuffer(44 + data);
  const v = new DataView(buf);
  const text = (at, s) => [...s].forEach((c, i) => v.setUint8(at + i, c.charCodeAt(0)));
  text(0, "RIFF");
  v.setUint32(4, 36 + data, true);
  text(8, "WAVE");
  text(12, "fmt ");
  v.setUint32(16, 16, true);
  v.setUint16(20, 1, true);
  v.setUint16(22, 1, true);
  v.setUint32(24, rate, true);
  v.setUint32(28, rate * 2, true);
  v.setUint16(32, 2, true);
  v.setUint16(34, 16, true);
  text(36, "data");
  v.setUint32(40, data, true);
  samples.forEach((s, i) => v.setInt16(44 + i * 2, s, true));
  return new Uint8Array(buf);
}

/** The "I heard you" sound as a `data:audio/wav` URI, for `new Audio()`. */
export function heardSoundUri() {
  const bytes = wavBytes(heardSoundSamples(), HEARD_SOUND.rate);
  let bin = "";
  bytes.forEach((b) => {
    bin += String.fromCharCode(b);
  });
  return `data:audio/wav;base64,${btoa(bin)}`;
}
